File: backends/fastapi/app/platform/token.py
from __future__ import annotations

from datetime import timedelta
from typing import Protocol

import jwt

from app.domain import errors
from app.domain.models import Role
from app.usecase.ports import Clock, IDGenerator


class RefreshStore(Protocol):
    async def save(self, jti: str, user_id: str, ttl: timedelta) -> None: ...
    # Atomic: returns the owner and removes the token in one step (Redis GETDEL).
    async def consume(self, jti: str) -> str | None: ...


class TokenService:
    """Short-lived access JWTs plus opaque, rotating refresh tokens. Access tokens are
    stateless HS256 JWTs; refresh tokens are server-side handles that are revoked on
    use, so a stolen-and-replayed refresh token fails. Same design as the Go backend.
    """

    def __init__(
        self,
        secret: str,
        access_ttl: timedelta,
        refresh_ttl: timedelta,
        store: RefreshStore,
        ids: IDGenerator,
        clock: Clock,
    ):
        self._secret = secret
        self._access_ttl = access_ttl
        self._refresh_ttl = refresh_ttl
        self._store = store
        self._ids = ids
        self._clock = clock

    def issue_access(self, user_id: str, role: Role) -> tuple[str, int]:
        now = self._clock.now()
        payload = {
            "sub": user_id,
            "role": role.value,
            "iat": int(now.timestamp()),
            "exp": int((now + self._access_ttl).timestamp()),
        }
        token = jwt.encode(payload, self._secret, algorithm="HS256")
        return token, int(self._access_ttl.total_seconds())

    async def issue_refresh(self, user_id: str) -> str:
        jti = self._ids.new_id()
        await self._store.save(jti, user_id, self._refresh_ttl)
        return jti

    async def rotate(self, refresh_token: str) -> str:
        user_id = await self._store.consume(refresh_token)
        if user_id is None:
            raise errors.INVALID_TOKEN  # unknown, expired, or already spent
        return user_id

    def parse_access(self, token: str) -> tuple[str, Role]:
        try:
            claims = jwt.decode(token, self._secret, algorithms=["HS256"])
        except jwt.PyJWTError as exc:
            raise errors.INVALID_TOKEN from exc
        sub = claims.get("sub")
        if not sub:
            raise errors.INVALID_TOKEN
        return sub, Role(claims.get("role", Role.CUSTOMER.value))

File: backends/express/src/platform/token.ts
import * as jwt from 'jsonwebtoken';

import { Errors } from '../domain/errors';
import { Role } from '../domain/models';
import { Clock, IdGenerator, TokenService } from '../usecase/ports';

// consume must be atomic: returns the owner and removes the token in one step (GETDEL).
export interface RefreshStore {
  save(jti: string, userId: string, ttlSeconds: number): Promise<void>;
  consume(jti: string): Promise<string | null>;
}

// Short-lived access JWTs plus opaque, rotating refresh tokens. Rotating one revokes
// it, so a stolen-and-replayed refresh token fails.
export class JwtTokenService implements TokenService {
  constructor(
    private readonly secret: string,
    private readonly accessTtlSeconds: number,
    private readonly refreshTtlSeconds: number,
    private readonly store: RefreshStore,
    private readonly ids: IdGenerator,
    private readonly clock: Clock,
  ) {}

  issueAccess(userId: string, role: Role): { token: string; expiresIn: number } {
    const now = Math.floor(this.clock.now().getTime() / 1000);
    const token = jwt.sign({ sub: userId, role, iat: now, exp: now + this.accessTtlSeconds }, this.secret, {
      algorithm: 'HS256',
    });
    return { token, expiresIn: this.accessTtlSeconds };
  }

  async issueRefresh(userId: string): Promise<string> {
    const jti = this.ids.newId();
    await this.store.save(jti, userId, this.refreshTtlSeconds);
    return jti;
  }

  async rotate(refreshToken: string): Promise<string> {
    const userId = await this.store.consume(refreshToken);
    if (!userId) throw Errors.InvalidToken;
    return userId;
  }

  parseAccess(token: string): { userId: string; role: Role } {
    let claims: jwt.JwtPayload;
    try {
      claims = jwt.verify(token, this.secret, { algorithms: ['HS256'] }) as jwt.JwtPayload;
    } catch {
      throw Errors.InvalidToken;
    }
    const sub = claims.sub;
    if (!sub || typeof sub !== 'string') throw Errors.InvalidToken;
    return { userId: sub, role: (claims.role as Role) ?? Role.Customer };
  }
}
